src/Symfony/Component/Validator/Constraints/BackedEnumValueValidator.php

```diff
@@ -0,0 +1,122 @@
+        } catch (\TypeError) {
+            $cases = $constraint->type::cases();
+            if (\count($cases) > 0) {
+                $this->context->buildViolation($constraint->message)
+                    ->setParameter('{{ value }}', $this->formatValue($value))
+                    ->setParameter('{{ choices }}', $this->formatValues(array_map(fn ($case) => $case->value, $cases)))
+                    ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                    ->addViolation();
+            }
+
+            return;
+        }
+
+        if (null === $enumTypeValue) {
+            $values = $constraint->type::cases();
+
+            if (\count($constraint->only) > 0) {
+                $values = $constraint->only;
+            } elseif (\count($constraint->except) > 0) {
+                $values = $constraint->type::cases();
+
+                foreach ($values as $name => $enumValue) {
+                    if (\in_array($enumValue, $constraint->except, strict: true)) {
+                        unset($values[$name]);
+                    }
+                }
+            }
+
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setParameter('{{ choices }}', $this->formatValues(array_map(fn ($case) => $case->value, $values)))
+                ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (\count($constraint->only) > 0 && !\in_array($enumTypeValue, $constraint->only, true)) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($enumTypeValue->value))
+                ->setParameter('{{ choices }}', $this->formatValues(array_map(fn ($case) => $case->value, $constraint->only)))
+                ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                ->addViolation();
+        }
+
+        if (\count($constraint->except) > 0 && \in_array($enumTypeValue, $constraint->except, true)) {
+            $values = $constraint->type::cases();
+
+            foreach ($values as $name => $enumValue) {
+                if (\in_array($enumValue, $constraint->except, strict: true)) {
+                    unset($values[$name]);
+                }
+            }
+
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($enumTypeValue->value))
+                ->setParameter('{{ choices }}', $this->formatValues(array_map(fn ($case) => $case->value, $values)))
+                ->setCode(BackedEnumValue::NO_SUCH_CHOICE_ERROR)
+                ->addViolation();
+        }
+    }
+
+    private function getNormalizedType(mixed $value): string
+    {
+        $type = \gettype($value);
+
+        if ('integer' === $type) {
+            return 'int';
+        }
+
+        return $type;
+    }
```

Comment:
This method feels unnecessarily generic, given that there are only two possible backing types.